### PR TITLE
Closes #2615: Adds titleView to DisplayToolbar.

### LINKED ISSUE #2615: Add a title view as optional text on the DisplayToolbar
In Fenix we will want the option to display the title of a website in custom tabs. Because of this, we'll need to create a `titleView` that is optionally shown in the `DisplayToolbar`

See [here](https://github.com/mozilla-mobile/fenix/issues/203#issuecomment-463426620) for more info.

### PR DESCRIPTION
**Pull request by sblatz**
### Pull Request checklist

- [x] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [x] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/android-components/blob/master/docs/changelog.md) or does not need one
- [x] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features

### CODE REVIEW

File: components/feature/customtabs/src/main/java/mozilla/components/feature/customtabs/CustomTabsToolbarFeature.kt [hunk 1]
```diff
@@ -23,6 +23,7 @@ import mozilla.components.concept.toolbar.Toolbar
 import mozilla.components.support.base.feature.BackHandler
 import mozilla.components.support.base.feature.LifecycleAwareFeature
 import mozilla.components.support.ktx.android.content.share
+import java.net.URL
```
Comment (jonalmeida) [inline components/feature/customtabs/src/main/java/mozilla/components/feature/customtabs/CustomTabsToolbarFeature.kt]: `android.net.Uri` or the extension function for `String.toUri` is probably better.

File: components/feature/toolbar/src/main/java/mozilla/components/feature/toolbar/ToolbarPresenter.kt [hunk 2]
```diff
@@ -10,6 +10,7 @@ import mozilla.components.browser.session.Session
 import mozilla.components.browser.session.SessionManager
 import mozilla.components.concept.toolbar.Toolbar
 import mozilla.components.feature.toolbar.internal.URLRenderer
+import java.net.URL
```
Comment (jonalmeida) [inline components/feature/toolbar/src/main/java/mozilla/components/feature/toolbar/ToolbarPresenter.kt]: `android.net.Uri` or the extension function for `String.toUri` here as well.

File: components/feature/toolbar/src/main/java/mozilla/components/feature/toolbar/ToolbarPresenter.kt [hunk 3]
```diff
@@ -64,16 +65,39 @@ class ToolbarPresenter(
...
     }
 
     override fun onUrlChanged(session: Session, url: String) {
-        renderer.post(url)
+        renderer.post(getURLToPost(session, url))
         toolbar.setSearchTerms(session.searchTerms)
     }
 
+    private fun getURLToPost(session: Session, url: String): String {
+        return if (session.isCustomTabSession()) {
```
Comment (jonalmeida) [inline components/feature/toolbar/src/main/java/mozilla/components/feature/toolbar/ToolbarPresenter.kt]: I'm not a fan of adding custom tab checks around the codebase since it increases our complexity in multiple components at once. How about always passing the url in, and when `displayTitle` is enabled, parse the url for the host in the CustomTabsToolbarFeature itself which does the global "isCustomTab" check in one place.

File: components/browser/toolbar/src/main/java/mozilla/components/browser/toolbar/BrowserToolbar.kt [hunk 4]
```diff
@@ -101,6 +101,12 @@ class BrowserToolbar @JvmOverloads constructor(
             }
         }
 
+    override var displayTitle: Boolean
+        get() = displayToolbar.titleView.isVisible()
+        set(value) {
+            displayToolbar.titleView.visibility = if (value) View.VISIBLE else View.GONE
+        }
```
Comment (pocmo) [inline components/browser/toolbar/src/main/java/mozilla/components/browser/toolbar/BrowserToolbar.kt]: Does this need to be a separate flag or can this be controlled by whether a title is set or not? Not every page has a title so I assume if there's no title then we do not want to display this view?

File: components/browser/toolbar/src/main/java/mozilla/components/browser/toolbar/display/DisplayToolbar.kt [hunk 5]
```diff
@@ -89,6 +89,15 @@ internal class DisplayToolbar(
         setOnClickListener(null)
     }
 
+    internal val titleView = AppCompatTextView(context).apply {
+        id = R.id.mozac_browser_toolbar_title_view
+        gravity = Gravity.CENTER_VERTICAL
+        textSize = URL_TEXT_SIZE
+        visibility = View.GONE
+
+        setSingleLine(true)
```
Comment (pocmo) [inline components/browser/toolbar/src/main/java/mozilla/components/browser/toolbar/display/DisplayToolbar.kt]: Do we want to define here what happens if the text is longer than the view's width?
-> https://developer.android.com/reference/android/widget/TextView.html#setEllipsize(android.text.TextUtils.TruncateAt)

File: components/feature/customtabs/src/main/java/mozilla/components/feature/customtabs/CustomTabsToolbarFeature.kt [hunk 6]
```diff
@@ -65,6 +66,12 @@ class CustomTabsToolbarFeature(
             if (config.showShareMenuItem) addShareButton(session)
             // Add menu items
             if (config.menuItems.isNotEmpty()) addMenuItems(config.menuItems)
+
+            toolbar.url = URL(session.url).host
```
Comment (pocmo) [inline components/feature/customtabs/src/main/java/mozilla/components/feature/customtabs/CustomTabsToolbarFeature.kt]: Not every URL has a host. Not every URL is http(s). :)

File: components/feature/toolbar/src/main/java/mozilla/components/feature/toolbar/ToolbarPresenter.kt [hunk 7]
```diff
@@ -64,16 +65,39 @@ class ToolbarPresenter(
...
+        val urlToPost = if (session != null) {
+            getURLToPost(session, session.url)
+        } else {
+            ""
+        }
+
+        // Set the title to be the url as we don't have a title until the page loads
+        session?.let { onTitleChanged(it, it.url) }
+
+        renderer.post(urlToPost)
```
Comment (pocmo) [inline components/feature/toolbar/src/main/java/mozilla/components/feature/toolbar/ToolbarPresenter.kt]: This logic shouldn't live here. The consuming app can already provide a `UrlRenderConfiguration` to define how the URL should be rendered. If we want a "host" only mode (Btw. is it really the host? Or the "registrable domain"?).

File: components/browser/toolbar/src/main/java/mozilla/components/browser/toolbar/display/DisplayToolbar.kt [hunk 8]
```diff
@@ -176,7 +185,13 @@ internal class DisplayToolbar(
 
     private val defaultColor = ContextCompat.getColor(context, R.color.photonWhite)
 
+    private var currentSiteSecurity = SiteSecurity.INSECURE
+
     internal var securityIconColor = Pair(defaultColor, defaultColor)
+        set(value) {
+            field = value
+            setSiteSecurity(currentSiteSecurity)
+        }
```
Comment (pocmo) [inline components/browser/toolbar/src/main/java/mozilla/components/browser/toolbar/display/DisplayToolbar.kt]: How is this change related? What's the background?

We try really hard to not have any state here and keep that in the `Session` only.

File: components/feature/customtabs/src/main/java/mozilla/components/feature/customtabs/CustomTabsToolbarFeature.kt [hunk 9]
```diff
@@ -65,6 +66,12 @@ class CustomTabsToolbarFeature(
...
+            toolbar.displayTitle = true
```
[the remark this answers is not in the document]
Comment (sblatz) [reply components/feature/customtabs/src/main/java/mozilla/components/feature/customtabs/CustomTabsToolbarFeature.kt]: Our UX asks for this to be [displayed universally](https://github.com/mozilla-mobile/fenix/issues/203#issuecomment-463426620) and doesn't make a mention of letting this be customizable. Are we sure we want to be setting it based on this flag?

File: components/browser/toolbar/src/main/java/mozilla/components/browser/toolbar/BrowserToolbar.kt (hunk 4 above)
Comment (sblatz) [reply components/browser/toolbar/src/main/java/mozilla/components/browser/toolbar/BrowserToolbar.kt]: Good point, I assumed every URL had a title.

File: components/browser/toolbar/src/main/java/mozilla/components/browser/toolbar/display/DisplayToolbar.kt (hunk 8 above)
Comment (sblatz) [reply components/browser/toolbar/src/main/java/mozilla/components/browser/toolbar/display/DisplayToolbar.kt]: I initially filed [a separate issue](https://github.com/mozilla-mobile/android-components/issues/2620) for that change, but wasn't sure if that was necessary. I'm happy to handle it in a separate PR :)

File: components/feature/toolbar/src/main/java/mozilla/components/feature/toolbar/ToolbarPresenter.kt (hunk 7 above)
Comment (sblatz) [reply components/feature/toolbar/src/main/java/mozilla/components/feature/toolbar/ToolbarPresenter.kt]: Our [mocks](https://github.com/mozilla-mobile/fenix/issues/203#issuecomment-463426620) used the word "host" and that seemed to give me the right string (aka "google.com"). How could I get the "registrable domain"?

Comment (pocmo) [reply components/feature/toolbar/src/main/java/mozilla/components/feature/toolbar/ToolbarPresenter.kt]: See `UrlRenderer` this is already used when coloring the domain. That's also a good example - if you look at desktop Firefox, the highlighted part of the URL is the "registrable domain".

File: components/browser/toolbar/src/main/java/mozilla/components/browser/toolbar/display/DisplayToolbar.kt (hunk 5 above)
Comment (sblatz) [reply components/browser/toolbar/src/main/java/mozilla/components/browser/toolbar/display/DisplayToolbar.kt]: Updated this, though I wonder why the `url` doesn't also have this set?

File: components/feature/customtabs/src/main/java/mozilla/components/feature/customtabs/CustomTabsToolbarFeature.kt (hunk 1 above)
Comment (sblatz) [reply components/feature/customtabs/src/main/java/mozilla/components/feature/customtabs/CustomTabsToolbarFeature.kt]: Removed both of these usages :)

File: components/feature/toolbar/src/main/java/mozilla/components/feature/toolbar/ToolbarPresenter.kt (hunk 3 above)
Comment (sblatz) [reply components/feature/toolbar/src/main/java/mozilla/components/feature/toolbar/ToolbarPresenter.kt]: If I go with [Sebastian's recommendation here](https://github.com/mozilla-mobile/android-components/pull/2625#discussion_r272075518) how should I handle this? I agree that constantly checking for `isCustomTab` could be annoying.

File: components/browser/toolbar/src/main/java/mozilla/components/browser/toolbar/display/DisplayToolbar.kt (hunk 5 above)
Comment (pocmo) [reply components/browser/toolbar/src/main/java/mozilla/components/browser/toolbar/display/DisplayToolbar.kt]: The url sets a fading edge in the `init {}` block.

File: components/feature/toolbar/src/main/java/mozilla/components/feature/toolbar/ToolbarPresenter.kt (hunk 3 above)
Comment (jonalmeida) [reply components/feature/toolbar/src/main/java/mozilla/components/feature/toolbar/ToolbarPresenter.kt]: This can still be done within the `CustomTabsToolbarFeature` using a session observer to check when the title is received in that session.

The [`WindowFeature`](https://github.com/mozilla-mobile/android-components/blob/master/components/feature/session/src/main/java/mozilla/components/feature/session/WindowFeature.kt#L19-L34) has a nice example of how to make an internal session observer in a feature class. However, since you want to focus on just one session, you wouldn't want to inherit from `SelectionAwareSessionObserver`, but provide the `sessionId` of the CT to it and make changes.

There might also be the case where the title is already received so you would have to check the `session.title` before deciding to not show the title entirely.

File: components/feature/customtabs/src/main/java/mozilla/components/feature/customtabs/CustomTabsToolbarFeature.kt (hunk 9 above)
[the remark this answers is not in the document]
Comment (jonalmeida) [reply components/feature/customtabs/src/main/java/mozilla/components/feature/customtabs/CustomTabsToolbarFeature.kt]: I've reached out to Bram offline to see if this was an oversight or if we want to intentionally not follow the Custom Tab spec for showing titles. I'll follow up here.

File: components/feature/toolbar/src/main/java/mozilla/components/feature/toolbar/ToolbarPresenter.kt (hunk 7 above)
Comment (sblatz) [reply components/feature/toolbar/src/main/java/mozilla/components/feature/toolbar/ToolbarPresenter.kt]: It looks like `CustomToolbarFeature` uses the same `oolbarPresenter` as `ToolbarFeature`. Because of this, I'm struggling to use a config that displays only the registrable domain. When the config is set to `showOnlyRegistrableDomain` for example, it will show in both custom and normal tabs, which is not the behavior we want.

Any recommendations on a clean way to solve this?

File: components/browser/toolbar/src/main/java/mozilla/components/browser/toolbar/display/DisplayToolbar.kt [hunk 10]
```diff
@@ -411,7 +437,21 @@ internal class DisplayToolbar(
 
         val iconWidth = if (siteSecurityIconView.isVisible()) siteSecurityIconView.measuredWidth else 0
         val urlLeft = navigationActionsWidth + iconWidth + urlBoxMargin
-        urlView.layout(urlLeft, 0, urlLeft + urlView.measuredWidth, measuredHeight)
+
```
[the remark this answers is not in the document]
Comment (sblatz) [reply components/browser/toolbar/src/main/java/mozilla/components/browser/toolbar/display/DisplayToolbar.kt]: Added a few more regarding layout!

File: components/feature/toolbar/src/main/java/mozilla/components/feature/toolbar/ToolbarPresenter.kt (hunk 3 above)
Comment (sblatz) [reply components/feature/toolbar/src/main/java/mozilla/components/feature/toolbar/ToolbarPresenter.kt]: I think my plan is to do something like mentioned in my discussion with Sebastian [below](https://github.com/mozilla-mobile/android-components/pull/2625#discussion_r272077771). Does that sound good to you? Just not sure what solution will work with the config being the same between the two bars.

**Comment by sblatz** [comment]
@jonalmeida, @pocmo: I believe I addressed all concerns raised. Please let me know if there's any other issues you see!

File: components/browser/toolbar/src/main/java/mozilla/components/browser/toolbar/display/DisplayToolbar.kt [hunk 11]
```diff
@@ -310,7 +330,14 @@ internal class DisplayToolbar(
         val urlWidth = (width - iconSize - browserActionsWidth - pageActionsWidth -
             menuWidth - navigationActionsWidth - 2 * urlBoxMargin)
         val urlWidthSpec = MeasureSpec.makeMeasureSpec(urlWidth, MeasureSpec.EXACTLY)
-        urlView.measure(urlWidthSpec, fixedHeightSpec)
+
+        titleView.measure(urlWidthSpec, fixedHeightSpec / MEASURED_HEIGHT_DENOMINATOR)
+
+        if (titleView.isVisible()) {
```
[the remark this answers is not in the document]
Comment (sblatz) [reply components/browser/toolbar/src/main/java/mozilla/components/browser/toolbar/display/DisplayToolbar.kt]: 👍 moved the titleView measured inside the if check

File: components/feature/toolbar/src/main/java/mozilla/components/feature/toolbar/ToolbarPresenter.kt [hunk 12]
```diff
@@ -64,7 +64,12 @@ class ToolbarPresenter(
         val session = sessionId?.let { sessionManager.findSessionById(sessionId) }
             ?: sessionManager.selectedSession
 
-        renderer.post(session?.url ?: "")
+        // Set the title to be the url as we don't have a title until the page loads
+        session?.let {
+            onTitleChanged(it, it.url)
```
[the remark this answers is not in the document]
Comment (sblatz) [reply components/feature/toolbar/src/main/java/mozilla/components/feature/toolbar/ToolbarPresenter.kt]: How should I handle a session **not** having a title until page load then? On desktop, we set the title to be the url until our observer gets notified. Is doing so not acceptable here?

File: components/feature/customtabs/src/main/java/mozilla/components/feature/customtabs/CustomTabsToolbarFeature.kt (hunk 9 above)
[the remark this answers is not in the document]
Comment (jonalmeida) [reply components/feature/customtabs/src/main/java/mozilla/components/feature/customtabs/CustomTabsToolbarFeature.kt]: We can land it this way for now and loop back when we hear back from @brampitoyo

**Comment by sblatz** [comment]
I have fixed all the issues raised. Unfortunately, now my tests are failing. For some reason whenever I call `updateTitle` the measuredHeight of both `urlTextViewRect` and `titleViewRect` are set to 0 (though their widths are correct.

@pocmo is there any chance you could take a look at my most recent commit to see if there is something you can figure out? I have tested this in sample browser and Fenix and the feature works as expected, so I think it's just a discrepancy in the test.

File: components/browser/toolbar/src/main/java/mozilla/components/browser/toolbar/display/DisplayToolbar.kt [hunk 13]
```diff
@@ -317,7 +337,15 @@ internal class DisplayToolbar(
         val urlWidth = (width - iconSize - browserActionsWidth - pageActionsWidth -
             menuWidth - navigationActionsWidth - 2 * urlBoxMargin)
         val urlWidthSpec = MeasureSpec.makeMeasureSpec(urlWidth, MeasureSpec.EXACTLY)
-        urlView.measure(urlWidthSpec, fixedHeightSpec)
+
+        if (titleView.isVisible()) {
+            // With a title view, the url and title split the rest of the space vertically
+            titleView.measure(urlWidthSpec, fixedHeightSpec / MEASURED_HEIGHT_DENOMINATOR)
+            urlView.measure(urlWidthSpec, fixedHeightSpec / MEASURED_HEIGHT_DENOMINATOR)
```
Comment (pocmo) [inline components/browser/toolbar/src/main/java/mozilla/components/browser/toolbar/display/DisplayToolbar.kt]: I haven't fully debugged it yet. But this is the part that looks wrong: `fixedHeightSpec` is a MeasureSpec (packed in an int) and not a height you can divide:
https://developer.android.com/reference/android/view/View.MeasureSpec

Instead you need to create a new spec (see where `fixedHeightSpec` gets created):
```Kotlin
 MeasureSpec.makeMeasureSpec(height / MEASURED_HEIGHT_DENOMINATOR /** Divide here! **/, MeasureSpec.EXACTLY)
```
